### PR TITLE
[BEAM-3345][SQL] Reject unsupported inputs into JOIN

### PR DESCRIPTION
**Pull request by akedin**
Allow only non-global windows with default trigger or end of window trigger.

[Joins Concerns Doc](https://docs.google.com/document/d/19Aw7YFqjwvL-2BD8M_ZAn2scChxc-_wWw6acTeM0RJ8/edit)

[Proposal Doc](https://docs.google.com/document/d/1V-ZgKVTwHdNSGlQWncWIzcf_Rw2oLKZFSkU43scLff4/edit#)

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by akedin** [comment]
R: @kennknowles does this make sense?

**Comment by akedin** [comment]
run gradle java precommit

**Comment by akedin** [comment]
run java gradle precommit

**Comment by akedin** [comment]
run java gradle precommit

**Comment by akedin** [comment]
run java gradle precommit

**Comment by akedin** [comment]
run java gradle precommit

**Comment by akedin** [comment]
run java gradle precommit

**Comment by akedin** [comment]
run java gradle precommit

**Comment by akedin** [comment]
run java gradle precommit

**Comment by akedin** [comment]
and it's green

**Comment by kennknowles** [comment]
Nice. I think this is really important.

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRel.java, line 213 at r1](https://reviewable.io:443/reviews/apache/beam/4642#-L4wbB3I-RJkArBkd8Wl:-L4wbB3I-RJkArBkd8Wm:b-e05f15) ([raw file](https://github.com/apache/beam/blob/eeba1dd4078f329e2b60e902f71c590c993df34a/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRel.java#L213)):*
> ```Java
>         && !triggersOncePerWindow(windowingStrategy)) {
>       throw new UnsupportedOperationException(
>           "Joining unbounded PCollections is currently only supported for "
> ```

Actually the name of the method would be a good first sentence here: "... is currently only supported for triggers that outputs once per window, such as the default trigger with zero allowed lateness."

Let's leave off "`AfterWatermark.pastEndOfWindow()` with any lateness" because that encourages users to use a kind of trigger that is confusing and drops data.

---

*[sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRel.java, line 228 at r1](https://reviewable.io:443/reviews/apache/beam/4642#-L4wba_bwmdi0F2NjrPq:-L4wba_bwmdi0F2NjrPr:bbsltu2) ([raw file](https://github.com/apache/beam/blob/eeba1dd4078f329e2b60e902f71c590c993df34a/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRel.java#L228)):*
> ```Java
>     }
> 
>     return AfterWatermark.pastEndOfWindow().equals(trigger)
> ```

Actually let's also reject this just because it is a bad idea to use it anyhow unless allowed lateness is zero (and then you might as well be using the default trigger). Does it come in as a continuation trigger or something?

---

*[sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlDslJoinTest.java, line 258 at r1](https://reviewable.io:443/reviews/apache/beam/4642#-L4wc-O08uJmU9R9QPXa:-L4wc-O08uJmU9R9QPXb:brmw946) ([raw file](https://github.com/apache/beam/blob/eeba1dd4078f329e2b60e902f71c590c993df34a/sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlDslJoinTest.java#L258)):*
> ```Java
>     PCollectionTuple inputs = tuple("ORDER_DETAILS1", orders, "ORDER_DETAILS2", orders);
> 
>     thrown.expectCause(isA(UnsupportedOperationException.class));
> ```

Let's add one or two `expectMessage` checks here to make sure the most important words or phrases show up. Not the whole error message because that test is too fragile. But just to make sure it isn't some unrelated exception masking it. Same below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4642#-:-L4wcDPwSbxl_IQ2vbWM:bwnxzec)*

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRel.java [hunk 1]
```diff
@@ -194,6 +203,28 @@ public BeamJoinRel(RelOptCluster cluster, RelTraitSet traits, RelNode left, RelN
...
+              + "In these cases Beam can guarantee it joins all input elements once per window. "
+              + windowingStrategy + " is not supported");
+    }
+  }
+
+  private boolean triggersOncePerWindow(WindowingStrategy windowingStrategy) {
+    Trigger trigger = windowingStrategy.getTrigger();
+
+    return !(windowingStrategy.getWindowFn() instanceof GlobalWindows)
+        && trigger instanceof DefaultTrigger
```
Comment (XuMingmin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamJoinRel.java]: From [JavaDoc](https://beam.apache.org/documentation/sdks/javadoc/2.2.0/org/apache/beam/sdk/transforms/windowing/DefaultTrigger.html), it's `equivalent to Repeatedly.forever(AfterWatermark.pastEndOfWindow())`, is it safe for `trigger instanceof DefaultTrigger` as I don't see a subclass for `DefaultTrigger`.

**Comment by kennknowles** [comment]
:lgtm: Once @XuMingmin is satisfied please squash commits for merge.

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4642#-:-L5KeZDetL1jouRfWiVy:bjfns0g)*

**Pull request merged**